### PR TITLE
Deprecate unified build in favor of unity build.

### PR DESCRIPTION
**Pull request by fruffy**
The custom unified build code might actually not be needed anymore. CMAKE has added [UNITY_BUILDS](https://cmake.org/cmake/help/latest/prop_tgt/UNITY_BUILD.html), which seem equivalent. 

I left the library in the cmake folder, in case downstream repositories still needed. There is only one build testing that the non-unity build compiles, which reduces the amount of CI runs. I can also make unity builds the default if requested.

### CODE REVIEW

**Comment by fruffy** [comment]
This requires fixes on some downstream repositories but otherwise works well. We may even drop the CI option to simplify tests.

**Comment by mbudiu-vmw** [comment]
Does this work with any version of cmake?
Is it faster or comparable to the old unified build?

**Comment by fruffy** [comment]
It works with CMake 3.16+, wich is the default since Ubuntu 20.04. The build times are roughly the same, but can be tweaked with `UNITY_BUILD_BATCH_SIZE`.

**Comment by mbudiu-vmw** [comment]
I personally don't have a problem with this, but I think this should work on as many platforms as possible.
If this precludes people running older versions of the software from building, then it could be problematic.

**Comment by fruffy** [comment]
I guess it boils down how we want to maintain backwards compatibility. We dropped support for anything lower than Ubuntu 20.04 with [this](https://github.com/p4lang/third-party/issues/25) pull request. So it might make sense to use features that are provided by it. 

The original unified_build library is still there. I can try to create a version that uses it depending on the CMake version.

**Comment by jfingerh** [comment]
"We dropped support for anything lower than Ubuntu 20.04 with https://github.com/p4lang/third-party/issues/25 pull request."

Techically, we stopped doing automated CI testing for anything lower than Ubuntu 20.04 with that change.

There was a commit in the last month or two that broke the build for Ubuntu 18.04, which when pointed out later via an issue then caused someone to create another PR that restored a working build for Ubuntu 18.04.

**Comment by jfingerh** [comment]
I agree that at some point it definitely makes sense not to support Ubuntu 18.04.  A reasonable question is: when?  One choice is right now.  Another would be to wait until 5 years after Ubuntu 18.04 was released in April 2023, when Ubuntu discontinues free support for it.

**Comment by fruffy** [comment]
The problem is that without CI we can not make any guarantees for 18.04. It will be quite hard to respect 18.04 restrictions going forward. This is related to the discussion in https://github.com/p4lang/p4c/issues/3451. Iirc that change worked because there was a solution that works for both 18.04 and future systems. This might not always be possible. 

For this particular PR, I can try to come up with a solution that can also work with older CMake systems.

**Comment by davidbolvansky** [comment]
> It works with CMake 3.16+, wich is the default since Ubuntu 20.04. The build times are roughly the same, but can be tweaked with `UNITY_BUILD_BATCH_SIZE`.

What is the plan now? Land this PR after April 2023? Or just land it now and one (on U18) can install newer cmake easily for example with pip install cmake==3.16.3.

**Comment by fruffy** [comment]
> What is the plan now? Land this PR after April 2023? Or just land it now and one (on U18) can install newer cmake easily for example with pip install cmake==3.16.3.

I will have to revive this PR but I am not quite sure what to do with `build_unified` yet. I will have to implement an equivalent version so that downstream projects can still use it. It might be nicer if that explicit function were not necessary.

**Comment by jfingerh** [comment]
I have no objections to merging updates that require specific later versions of cmake on Ubuntu 18.04, as long as there are clear README directions on building on Ubuntu 18.04 somewhere.

**Comment by davidbolvansky** [comment]
> build_unified

I would remove it. 

Could you create new cmake option so we could try new cmake native unified build downstream and do necessary changes?

File: CMakeLists.txt
Comment (davidbolvansky) [inline CMakeLists.txt:92]: if (ENABLE_UNIFIED_COMPILATION)
set(CMAKE_UNITY_BUILD ON)

File: backends/p4tools/common/lib/timer.cpp
Comment (fruffy) [reply backends/p4tools/common/lib/timer.cpp:91]: https://github.com/p4lang/p4c/pull/3785

File: CMakeLists.txt
Comment (davidbolvansky) [inline CMakeLists.txt:96]: Feel free to emit an warning here that this option is deprecated and it will be removed after some time.

**Review by davidbolvansky — COMMENTED** [review]
cmake_minimum_required / README  needs to be updated

File: .github/workflows/ci-p4tools.yml
Comment (rst0git) [inline .github/workflows/ci-p4tools.yml:35]: It looks like this pull request may have merge conflicts with https://github.com/p4lang/p4c/pull/3770
Should we merge the tests for Ubuntu 20.04 and 22.04 first?

Comment (davidbolvansky) [reply .github/workflows/ci-p4tools.yml:35]: Pushed. Please rebase.

**Comment by davidbolvansky** [comment]
For now please pip install cmake 3.16.3 here:
https://github.com/p4lang/p4c/blob/main/tools/ci-build.sh#L98

File: tools/ci-build.sh
Comment (rst0git) [inline tools/ci-build.sh:114]: By default the Ubuntu 18.04 container image does not have `cmake` installed. It might be better to remove `cmake` from `P4C_DEPS` above and use "if / else" statement like:
```bash
if [[ "${DISTRIB_RELEASE}" == "18.04" ]] ; then
    # For Ubuntu 18.04, remove the Debian package and install the pypi-supplied version instead.
    P4C_PIP_PACKAGES+=" cmake==3.16.3"
else
    P4C_DEPS+=" cmake"
fi
```

File: .github/workflows/ci-validation.yml
Comment (davidbolvansky) [inline .github/workflows/ci-validation.yml:43]: Are you sure? Why docker again?

Comment (fruffy) [reply .github/workflows/ci-validation.yml:43]: Something broke with this file while rebasing. Will fix.

File: .github/workflows/ci-validation.yml
Comment (davidbolvansky) [inline .github/workflows/ci-validation.yml:44]: You need this workdir

File: .github/workflows/ci-validation.yml
Comment (davidbolvansky) [inline .github/workflows/ci-validation.yml:44]: Will it run? dont you need sudo -E?

i am suprised why you need any changes at all in this file

Comment (davidbolvansky) [reply .github/workflows/ci-validation.yml:44]: Some hours ago when I was looking why validate fails it was due to some leftover of removed build_unified.

Comment (davidbolvansky) [reply .github/workflows/ci-validation.yml:44]: Try to avoid any changes here and see what CI says

Comment (fruffy) [reply .github/workflows/ci-validation.yml:44]: Yes it should run without sudo. `sudo` is only required for eBPF kernel tests.  These changes need to be preserved to avoid build failures of Gauntlet affecting PRs. This is because Gauntlet is a downstream dependency.

Comment (davidbolvansky) [reply .github/workflows/ci-validation.yml:44]: But this is still broken. Validate Step ends in Build failed (but build went fine, test image was created so strange…. what is worse, it is green, so easy to overlook any issues - I changed it in my previous PR but I am not sure why you reverted those changes).

File: .github/workflows/ci-test.yml
Comment (davidbolvansky) [inline .github/workflows/ci-test.yml:205]: GMP is removed

File: .github/workflows/ci-validation.yml
Comment (fruffy) [reply .github/workflows/ci-validation.yml:45]: Yes, that is intentional. Failed builds should not show up in PRs since they can only be fixed in the downstream repository. 

I will fix the odd skipping.

File: .github/workflows/ci-validation.yml
Comment (davidbolvansky) [reply .github/workflows/ci-validation.yml:40]: Removed subshell will break this line as you will not copy to itself (see cd above!)

File: .github/workflows/ci-validation.yml
Comment (davidbolvansky) [inline .github/workflows/ci-validation.yml:23]: I would default to false

File: .github/workflows/ci-validation.yml
Comment (davidbolvansky) [inline .github/workflows/ci-validation.yml:39]: (cd /p4c/ && sudo -E tools/ci-build.sh && echo "BUILD_SUCCESS=true" >> $GITHUB_ENV)

Comment (fruffy) [reply .github/workflows/ci-validation.yml:39]: Will wait for #3792.

**Comment by davidbolvansky** [comment]
Please fix failing CI - otherwise I see no blockers to proceed with this PR, do you?

File: .github/workflows/ci-validation.yml
Comment (vlstill) [inline .github/workflows/ci-validation.yml:38]: Why is it not sufficient to fail normally?

Comment (fruffy) [reply .github/workflows/ci-validation.yml:38]: This tool is not part of the P4C repository, if it does not build it should not block a PR. It must be fixed separately.

**Comment by fruffy** [comment]
@mbudiu-vmw @vhavel This PR needs authorized approval because it touches some Testgen code.

**Comment by mbudiu-vmw** [comment]
CI and cmake are not my speciality, so I'd rather have someone else comment.

**Review by davidbolvansky — APPROVED** [review]

**Comment by fruffy** [comment]
> CI and cmake are not my speciality, so I'd rather have someone else comment.

The issue is that this PR can not be emerged unless someone of the code owners approves. It is a little unfortunate.

**Comment by fruffy** [comment]
@vhavel @pkotikal This PR is stuck because of the reviewer lock we have on testgen files. If you have some time could you give this a review? This PR deprecates our homecooked unified build in favor of CMake's unity builds.

File: frontends/common/parser_options.cpp
Comment (vlstill) [inline frontends/common/parser_options.cpp:495]: As far as I understand it, the original code was actually using placeholders the way they were intended. There is basically nothing wrong with *local* `using namespace`.

File: .github/workflows/ci-validation.yml
Comment (fruffy) [reply .github/workflows/ci-validation.yml:38]: I believe most of the tests being non-required is an accident. We do not have control over adjusting the settings which of the tests are required. We should fix this. 

I am okay with removing this check, I maintain the tool after all. 

@mbudiu-vmw Should we make this a non-required test?

File: frontends/common/parser_options.cpp
Comment (fruffy) [reply frontends/common/parser_options.cpp:495]: For bf-p4c, I had to make the namespace explicit because the unified builds also ended up including boost placeholders. This will lead to a conflict. It seems to be an unfortunate side-effect of including boost.

File: backends/p4tools/modules/testgen/targets/bmv2/backend/stf/stf.cpp
Comment (mbudiu-vmw) [inline backends/p4tools/modules/testgen/targets/bmv2/backend/stf/stf.cpp:238]: how are these changes related to unified builds?

**Review by mbudiu-vmw — COMMENTED** [review]
This looks fine, but I really don't know whether the new requirements will prevent people with older software from building the compiler.

File: backends/p4tools/modules/testgen/targets/bmv2/backend/stf/stf.cpp
Comment (fruffy) [reply backends/p4tools/modules/testgen/targets/bmv2/backend/stf/stf.cpp:238]: When multiple `.cpp` files are compiled together this function violates the one-definition rule and the compiler complains.

**Comment by fruffy** [comment]
> This looks fine, but I really don't know whether the new requirements will prevent people with older software from building the compiler.

We do have a CI run for Ubuntu 18, which already requires some workarounds. Ubuntu 16.04 we dropped support a long time ago. I do not know about other systems.

File: frontends/common/parser_options.cpp
Comment (vlstill) [reply frontends/common/parser_options.cpp:495]: Maybe someone is using `using namespace boost::placeholders` globally / on .cpp level? That would be the actual bug... (or not exactly bug, but in my opinion that would be the wrong use of the using in presence of something like unity build that breaks file scopes).

Comment (fruffy) [reply frontends/common/parser_options.cpp:495]: I believe [this](https://www.boost.org/doc/libs/1_71_0/boost/bind.hpp) file is getting pulled in somewhere. 

Related: 
https://stackoverflow.com/questions/53203970/why-boostbind-insists-pulling-boostplaceholders-into-global-namespace

Comment (vlstill) [reply frontends/common/parser_options.cpp]: We have discussed this internally, the problem is some other boost headers pull the `using namespace boost::placeholders` into global namespace (at least in older boost versions). Therefore this cannot be fixed nicely. In order to allow inclusion of such boost headers (e.g. `<boost/graph/transitive_closure.hpp>` for boost versions < 1.77) we need to keep the uses of `std::placeholders::_1` fully qualified). Thanks @fruffy for tracking the source.

**Review by vlstill — APPROVED** [review]

**Comment by fruffy** [comment]
@mbudiu-vmw  @vlstill  and @davidbolvansky  both approved but I can not merge this PR since it is missing an authoritative approval.

**Review by jnfoster — APPROVED** [review]

**Comment by jnfoster** [comment]
I believe I have the power. Done.

**Comment by fruffy** [comment]
Thank you!

**Pull request merged**
